src/irmin-pack/unix/file_manager.ml

```diff
@@ -0,0 +1,365 @@
+
+  (** Note on SWMR consistency: It is undefined for a reader to attempt an
+      opening before [create_rw] is over.
+
+      Note on crash consistency: Crashing during [create_rw] leaves the storage
+      in an undefined state.
+
+      Note on errors: If [create_rw] returns an error, the storage is left in an
+      undefined state and some file descriptors might not be closed. *)
+  let create_rw ~overwrite config =
+    let open Result_syntax in
+    let root = Irmin_pack.Conf.root config in
+    let* () =
+      match (overwrite, Io.classify_path root) with
+      | _, `File -> Error `Not_a_directory
+      | false, `Directory -> Error `File_exists
+      | true, `Directory -> Ok ()
+      | _, `No_such_file_or_directory -> Io.mkdir root
+    in
+    let* control =
+      let status =
+        Payload.From_v3_gc_allowed
+          { entry_offset_suffix_start = Int63.zero; gc_generation = 0 }
+      in
+      create_control_file ~overwrite config status
+    in
+    let make_dict = Dict.create_rw ~overwrite in
+    let make_suffix = Suffix.create_rw ~overwrite in
+    let make_index ~flush_callback ~readonly ~throttle ~log_size root =
+      (* [overwrite] is ignored for index *)
+      Index.v ~fresh:true ~flush_callback ~readonly ~throttle ~log_size root
+    in
+    finish_constructing_rw config control ~make_dict ~make_suffix ~make_index
+
+  (* Open rw **************************************************************** *)
+
+  let open_rw_with_control_file config =
+    let open Result_syntax in
+    let root = Irmin_pack.Conf.root config in
+    let* control =
+      let path = Irmin_pack.Layout.V3.control ~root in
+      Control.open_rw ~path
+    in
+    let pl : Payload.t = Control.payload control in
+    let dead_header_size =
+      match pl.status with
+      | From_v1_v2_post_upgrade _ -> legacy_io_header_size
+      | From_v3_gc_disallowed | From_v3_gc_allowed _ -> 0
+    in
+    let make_dict =
+      let end_offset = pl.dict_offset_end in
+      Dict.open_rw ~end_offset ~dead_header_size
+    in
+    let make_suffix =
+      let end_offset = pl.entry_offset_suffix_end in
+      Suffix.open_rw ~end_offset ~dead_header_size
+    in
+    let make_index ~flush_callback ~readonly ~throttle ~log_size root =
+      Index.v ~fresh:false ~flush_callback ~readonly ~throttle ~log_size root
+    in
+    finish_constructing_rw config control ~make_dict ~make_suffix ~make_index
+
+  let open_rw_migrate_from_v1_v2 config =
+    let open Result_syntax in
+    let root = Irmin_pack.Conf.root config in
+    let src = Irmin_pack.Layout.V1_and_v2.pack ~root in
+    let dst = Irmin_pack.Layout.V3.suffix ~root ~generation:0 in
+    let* () = Io.move_file ~src ~dst in
+    let* entry_offset_at_upgrade_to_v3 =
+      let* suffix = Io.open_ ~path:dst ~readonly:true in
+      let* x = Io.read_size suffix in
+      let+ () = Io.close suffix in
+      let ( - ) = Int63.sub in
+      x - Int63.of_int legacy_io_header_size
```

Comment:
Let's read that info from the header instead of guessing it using `Io.read_size`